samcli/hook_packages/terraform/hooks/prepare/enrich.py

```diff
@@ -455,7 +457,16 @@ def _get_source_code_path(
                 f"should contain the lambda function/lambda layer "
                 f"{src_code_attribute_name} in property {src_code_property_name}, and it should not be an empty list"
             )
-        cfn_source_code_path = source_code[0]
+        modified_sam_metadata_resource = sam_metadata_resource
```

Comment:
My intention was to create a copy of the metadata object to avoid modifying the original and causing unintended changes elsewhere. I now know that my approach was incorrect after I read a bit about how copying works in Python.  I think what I wanted to do was to a deep copy. 

I'd appreciate some feedback about what the right choice is. I could change it to a deep copy, remove the shallow copy since the behavior should be the same with or without it, or rework the logic flow to do the "if list" check before the if dict/if string checks

Reply:
> or rework the logic flow to do the "if list" check before the if dict/if string checks

Would that get rid of the recursive call of `_get_source_code_path` on L462? It seems it won't cause any issue but just want to see if we can make the code look cleaner.

Otherwise, I think just going with deep copy is fine.